### ISSUE TITLE
[Feature]: Depend on Test, function or Group

### ISSUE THREAD

**Issue by srireddy4** [microsoft/playwright#29530]
### 🚀 Feature Request

When orchestrating tests within a class, I opt to begin by offering the choice of running the initial two tests either concurrently or sequentially and the order of tests preview as well. Once this initial pair is completed, the subsequent tests are queued to await their turn. This method affords the flexibility to trigger the next pair of tests based on the outcome—whether successful or unsuccessful— giving the option either to run on fail or not. This cycle repeats iteratively until all tests have been executed.

### Example

Write in Java as I took this from the inspiration of TestNG Java
@Test
public void test1(){}

@Test
public void test2(){}

@Test(dependOnMethods ={"test1","test2"})
public void test3(){}

@Test(dependOnMethods ={"test1","test2"})
public void test4(){}

repeat the cycle

### Motivation

I drew inspiration from TestNG annotations for this approach. While working with numerous clients across various applications and integration scenarios, I often encountered the need to control the priority order of tests, whether executing them concurrently or sequentially. Despite the recommended practice of building independent tests, there were situations where I found it necessary to utilize functionalities like 'dependsOnGroups' or 'dependsOnMethods' from TestNG. Even as I transition towards adopting Playwright, a tool I see as the future, I've noticed a gap in a feature that has been common throughout my career. TestNG's widespread use and its annotations have been instrumental. As I embrace Playwright, I anticipate a similar need for a feature akin to the ones offered by TestNG for managing test execution.

**Comment by yury-s** [comment]
I believe [project dependencies](https://playwright.dev/docs/test-global-setup-teardown#option-1-project-dependencies) already meet your requirements, is there anything you cannot achieve with them?

**Comment by srireddy4** [comment]
@yury-s The project dependencies fall short of meeting my needs, as demonstrated by the examples provided. I'm referring to scenarios where, for instance, out of 10 tests, tests 1 and 2 run concurrently, while test 3 waits until tests 1 and 2 complete and runs sequentially. Then, tests 4 and 5 run concurrently upon completion of test 3, and this sequence continues until all 10 test cases are finished. This complicated series of dependencies doesn't work using project dependencies, As project 
 dependencies simply running a couple of tests before executing another. If it's possible to achieve this using project dependencies, could you please provide an example in Playwright? Thnks.!

**Comment by PsiKai** [comment]
I am also looking for a feature similar to project dependencies, but rather, on a `test.describe` block basis.

Example:

- Describe block contains 10 individual tests
- One test will run before to execute setup for the 10 tests
- After the setup, all 10 tests can be executed fully parallel

With serial mode and the setup in a `beforeAll` block, one failing test causes all others to not be tried.  In parallel mode, the `beforeAll` setup is executed per worker, and is unnecessary work in my CI container.

These tests are already part of a bigger project (and are run across different projects), so it would not make sense to put them into their own project with their own dependency.  With project dependencies it is _technically_ possible, but at the expense of additional cognitive load and endless cli flags to properly coordinate them all.

Thanks for your consideration!

**Comment by srireddy4** [comment]
@yury-s, Thank you for sharing the example. While it partially addresses our needs, implementing this approach at the project level poses an inconvenience. Projects typically involve higher-level complexities with specific dependencies tailored to individual project requirements. Placing each functional test in the project's configuration level solely for handling dependencies doesn't seem practical. Rarely does it prove advantageous to manage such aspects within the confines of tests.describe or test.describe.configure. Hundreds or thousands of tests can be constructed using this passive approach, which would significantly aid in the tests.describe section, rather than in projects.

**Comment by pavelfeldman** [comment]
#### Why was this issue closed?

Thank you for your involvement. This issue was closed due to limited engagement (upvotes/activity), lack of recent activity, and insufficient actionability. To maintain a manageable database, we prioritize issues based on these factors.

If you disagree with this closure, please open a new issue and reference this one. More support or clarity on its necessity may prompt a review. Your understanding and cooperation are appreciated.
